### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by springmeyer** [mapbox/mapbox-studio-classic#315]
TM2 does not work on windows yet, but it should be feasible to port it and have it work reasonably well in a modern browser like Chrome.

This ticket stands to track this task.

Porting will require:
- [x] - setting `process.env.HOME = process.env.USERPROFILE;` at startup (https://github.com/mapbox/tm2/commit/019471c42b9dd45be582223b290b4690f85f69d3#diff-168726dbe96b3ce427e7fedce31bb0bc)
- [x] - Fixing `EPERM, open` error when trying to open `app.db`. The problem is that the deletion here https://github.com/mapbox/tm2/blob/23196a54a6d4d3f46b6e8847c437cbeca2eb4d68/lib/tm.js#L46 is not possible on windows because the file handle is still open.
- [x] - Fixing broken file browser: `EPERM: stat` error when trying to list `C:\`
- [x] - Fixing invalid path completion leading to `UNKNOWN, readdir`: somewhere a path is ending up like `\\C:\users\<username` due to the `/` prefixing right here: https://github.com/mapbox/tm2/blob/23196a54a6d4d3f46b6e8847c437cbeca2eb4d68/index.js#L306
- [x] - ~~Fix broken source/style.tm2 saving that leads to paths like `http://localhost:3000/style?id=tmstyle://C:\Users\mapnik/foo.tm2`~~ Not fixed (still odd looking path) but the odd looking path appears to be harmless for saving and opening, which now works.
- [x] node-blend is not windows-ready atm. Possibly moving away from it to node-mapnik for image stitching. https://github.com/mapbox/node-blend/issues/48

**Comment by springmeyer** [comment]
Updated the list of broken things above. We're not very far off from supporting windows. The problems are restricted to filesystem stuff. By working around the `app.db` problem I got the default project to load:

![screenshot 2014-04-25 14 17 38](https://cloud.githubusercontent.com/assets/20300/2805863/4f0a0364-ccbf-11e3-9ce4-b4481a24b53b.png)

**Comment by yhahn** [comment]
Above commit should (?) fix no 2 `EPERM` issue.

**Comment by springmeyer** [comment]
EPERM issue is fixed, thanks! Tested on windows and updated the above checkboxes to reflect that saving styles also works now. File browser is still broken (does not display any entries and does not allow moving up a directory) so I can't yet test source creation:

**Comment by yhahn** [comment]
Word, will consider that next up.

**Comment by springmeyer** [comment]
seeing this when I try to open the browser and click up a directory:
![screenshot 2014-05-13 13 33 53](https://cloud.githubusercontent.com/assets/20300/2963685/0c2229f0-dade-11e3-82df-520b272d8354.png)

**Comment by wilhelmberg** [comment]
`EPERM, open` still exists on Windows when `app.db` does **not** exist.
After I created an empty file tm2 was able to start and load the default project.

**Comment by springmeyer** [comment]
@BergWerkGIS okay, perhaps a `fs.exists` check could be added to https://github.com/mapbox/tm2/blob/master/lib/tm.js#L81-L96 to avoid that first run problem. Want to give that a shot?

**Comment by wilhelmberg** [comment]
Sure.

**Comment by wilhelmberg** [comment]
@springmeyer 
Hmh, probably I'm missing some context here, because some strange things are going on that I don't fully understand.

tl;dr:
Inserting `if (!existsSync(filepath)) { fs.writeFileSync(filepath, ""); }`
as the first line in [dbcompact](https://github.com/mapbox/tm2/blob/master/lib/tm.js#L82) solves the problem.
Seems like a dirty hack to me, though.

---

If the `app.db` does not exist it is created via [`var old = dirty(filepath);`](https://github.com/mapbox/tm2/blob/master/lib/tm.js#L82) and is available to unlink, which succeeds.
Nevertheless creating the new db with [`var db = dirty(filepath);`](https://github.com/mapbox/tm2/blob/master/lib/tm.js#L87) doesn't seem to be successful, because at entry of [dbmigrate](https://github.com/mapbox/tm2/blob/master/lib/tm.js#L56) it is not available.
Neither is it possible to create a new file at that code location (EPERM).
The error thrown `EPERM, open` seems to come from another location, because `db.get('version')` in `dbmigrate` returns `undefined` and the `switch` isn't executed.

On the other hand, if `app.db` exists (even if it is an empty file created with the patch mentioned above) everything works as expected and I'm shown the `Connect your Mapbox account` page.

**Comment by yhahn** [comment]
@BergWerkGIS I pushed an adjustment based on your notes to the `dbcompact-winfix` branch -- the idea being to not attempt to do db compaction (e.g. reading the db into memory and then unlinking the stale file) if the db doesn't exist in the first place.

Would love to know how it does for you -- I'll try to get tm2 up and running on a VM sometime today so I can debug more directly.

**Comment by wilhelmberg** [comment]
@yhahn cannot test your adjustment: I was testing with an older commit (unfortunately don't remember which one exactly), because latest source fails at installing `node-blend`.
![tm2-blend](https://cloud.githubusercontent.com/assets/4549888/3236640/0a8d7e74-f0e0-11e3-83d9-7c8206027ccb.png)

**Comment by wilhelmberg** [comment]
@yhahn I saw that your latest commit worked on [AppVeyor](https://ci.appveyor.com/project/springmeyer/tm2/build/1.0.190), so I tested on two different machines with node 0.10.26, 0.10.28 and 0.10.29.

I always deleted the tm2 folder completely and started from scratch.

With 26/28 I get varying results: sometimes `node-srs` gets rebuilt, sometimes not.
Sometimes it fails like in the screenshot of my previous comment.

With 29 nothing gets compiled, but this error is shown:
![tm2-abaculus](https://cloud.githubusercontent.com/assets/4549888/3240818/4d76256c-f133-11e3-9c9c-113d171cf4db.png)

A manual `npm install abacus` produces the same error as in the screenshot of my previous comment.

**Comment by wilhelmberg** [comment]
@yhahn I tried to get `pkg-config` working as this what abacus seems to be looking for.
I used precompiled files from here: http://www.gtk.org/download/win32.php
Error now is `No package 'libpng' found`.
Any ideas?
![abacus-pkgconfig](https://cloud.githubusercontent.com/assets/4549888/3241093/744d3856-f138-11e3-8770-e9d10316ba65.png)

**Comment by wilhelmberg** [comment]
@springmeyer @yhahn 

Ok, worked my way thru the commits.

This is the last one that is working for me: [Merge pull request #414 from mapbox/tv102](https://github.com/mapbox/tm2/commit/2d63d939ed8cebeda08367275c7d0d278a302b0e).

The next one ([abaculus updated for headers](https://github.com/mapbox/tm2/commit/e609ffab57683b870f70237d37f747d89c035a46)) isn't.
It fails with the `pkg-config` error.

**Comment by yhahn** [comment]
@BergWerkGIS yeah, for now I've removed the `abaculus` dependency from https://github.com/mapbox/tm2/tree/dbcompact-winfix and am focusing on getting the tests pass without it. Not trying to even run the app yet : )

@springmeyer is looking at https://github.com/mapbox/node-blend/issues/48 so once that is resolved we can re-add the dep.

**Comment by wilhelmberg** [comment]
@yhahn Ok, wasn't aware about that.

Just to let you know how the latest commit looks like on my main machine.
`npm install` is only successfull with node 0.10.29 32bit. Previous versions fail.

![tm2-npm-test-01](https://cloud.githubusercontent.com/assets/4549888/3255790/46226528-f20f-11e3-9e02-cc959a6bf889.png)

![tm2-npm-test-02](https://cloud.githubusercontent.com/assets/4549888/3255792/4ba7d1b8-f20f-11e3-8cce-ef05eb2a7f70.png)

**Comment by springmeyer** [comment]
after #457 appveyor builds are now showing errors in the logs (https://ci.appveyor.com/project/Mapbox/tm2/build/1.0.4#L990) - which look similar to @BergWerkGIS's logs above.

**Comment by yhahn** [comment]
https://github.com/mapbox/tm2/pull/468

appveyor builds are now completing and passing. This gets us CI for the server component of tm2 -- next up for me is seeing if we can get our browser tests running + passing on windows.

**Comment by yhahn** [comment]
https://github.com/mapbox/tm2/pull/469

client-side testing in place. I haven't actually used the app on windows much and I see that as the next step as our tests are far from comprehensive.

**Comment by wilhelmberg** [comment]
This is, what it looks like on my machine (cannot add any data :sob: ):
![tm2](https://cloud.githubusercontent.com/assets/4549888/3379227/0b1167fc-fbee-11e3-92ad-08d523f6ea3d.png)

**Comment by yhahn** [comment]
Closing this out in favor of more specific tickets now.
